sqlmesh/core/context.py

```diff
@@ -2287,22 +2290,45 @@ def _context_diff(
             ensure_finalized_snapshots=ensure_finalized_snapshots,
             diff_rendered=diff_rendered,
             environment_statements=self._environment_statements,
+            gateway_managed_virtual_layer=self.gateway_managed_virtual_layer,
         )
 
     def _run_janitor(self, ignore_ttl: bool = False) -> None:
+        # Clean up expired environments by removing their views and schemas
         self._cleanup_environments()
-        expired_snapshots = self.state_sync.delete_expired_snapshots(ignore_ttl=ignore_ttl)
+
+        # Identify and delete expired snapshots
+        cleanup_targets = self.state_sync.delete_expired_snapshots(ignore_ttl=ignore_ttl)
+
+        # Remove the expired snapshots tables
         self.snapshot_evaluator.cleanup(
-            expired_snapshots,
-            self._snapshot_gateways,
+            target_snapshots=cleanup_targets,
             on_complete=self.console.update_cleanup_progress,
         )
 
         self.state_sync.compact_intervals()
 
     def _cleanup_environments(self) -> None:
         expired_environments = self.state_sync.delete_expired_environments()
-        cleanup_expired_views(self.engine_adapter, expired_environments, console=self.console)
+
+        environment_snapshot_adapters: t.Dict[str, t.Dict[str, EngineAdapter]] = {}
+        for environment in expired_environments:
+            snapshot_adapters: t.Dict[str, EngineAdapter] = {}
+            if environment.gateway_managed_virtual_layer:
+                snapshots = self.state_sync.get_snapshots(environment.snapshots)
+                for snapshot_id, snapshot in snapshots.items():
+                    if snapshot.is_model and not snapshot.is_symbolic:
+                        snapshot_adapters[snapshot_id.name] = self._get_engine_adapter(
```

Comment:
If we included `model_gateway` into `SnapshotTableInfo` we wouldn't have to do this.